presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PartialAggregationPushDown.java

```diff
@@ -79,17 +77,114 @@ public Rewriter(SymbolAllocator allocator, PlanNodeIdAllocator idAllocator)
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> context)
         {
+            PlanNode child = node.getSource();
+
+            if (!(child instanceof ExchangeNode)) {
+                return context.defaultRewrite(node);
+            }
+
+            // partial aggregation can only be pushed through exchange that doesn't change
+            // the cardinality of the stream (i.e., gather or repartition)
+            ExchangeNode exchange = (ExchangeNode) child;
+            if (exchange.getType() != GATHER && exchange.getType() != REPARTITION) {
+                return context.defaultRewrite(node);
+            }
+
+            if (exchange.getType() == REPARTITION) {
+                if (exchange.getPartitioningScheme().isReplicateNulls()) {
+                    return context.defaultRewrite(node);
+                }
+
+                // if partitioning columns are not a subset of grouping keys,
+                // we can't push this through
+                List<Symbol> partitioningColumns = exchange.getPartitioningScheme()
+                        .getPartitioning()
+                        .getArguments()
+                        .stream()
+                        .map(Partitioning.ArgumentBinding::getColumn)
+                        .collect(Collectors.toList());
+
+                if (!node.getGroupingKeys().containsAll(partitioningColumns)) {
+                    return context.defaultRewrite(node);
+                }
+            }
+
+            // currently, we only support plans that don't use pre-computed hash functions
+            if (node.getHashSymbol().isPresent() || exchange.getPartitioningScheme().getHashColumn().isPresent()) {
+                return context.defaultRewrite(node);
+            }
+
             boolean decomposable = node.getFunctions().values().stream()
                     .map(functionRegistry::getAggregateFunctionImplementation)
                     .allMatch(InternalAggregationFunction::isDecomposable);
-            checkState(node.getStep() == SINGLE, "aggregation should be SINGLE, but it is %s", node.getStep());
-            checkState(context.get() == null, "context is not null: %s", context);
-            if (!decomposable || !allowPushThrough(node.getSource())) {
+
+            if (!decomposable) {
                 return context.defaultRewrite(node);
             }
 
+            if (node.getStep() == SINGLE) {
+                return context.rewrite(split(node));
+            }
+            else if (node.getStep() == PARTIAL) {
+                return context.rewrite(pushPartial(node, exchange));
+            }
+
+            return context.defaultRewrite(node);
+        }
+
+        private PlanNode pushPartial(AggregationNode aggregation, ExchangeNode exchange)
+        {
+            ImmutableList.Builder<List<Symbol>> exchangeAssignments = ImmutableList.builder();
+            ImmutableList.Builder<PlanNode> partials = ImmutableList.builder();
```

Comment:
If you make the changes I suggest below, the code will be easier if this is a regular `List`